collect/central_collector.go

```diff
@@ -0,0 +1,250 @@
+package collect
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"runtime/debug"
+	"sync"
+	"time"
+
+	"github.com/honeycombio/refinery/centralstore"
+	"github.com/honeycombio/refinery/collect/cache"
+	"github.com/honeycombio/refinery/config"
+	"github.com/honeycombio/refinery/logger"
+	"github.com/honeycombio/refinery/metrics"
+	"github.com/honeycombio/refinery/sample"
+	"github.com/honeycombio/refinery/transmit"
+	"github.com/honeycombio/refinery/types"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	cacheEjectBatchSize = 100
+)
+
+type CentralCollector struct {
+	Store          centralstore.SmartStorer `inject:""`
+	Config         config.Config            `inject:""`
+	Clock          clockwork.Clock          `inject:""`
+	Transmission   transmit.Transmission    `inject:"upstreamTransmission"`
+	Logger         logger.Logger            `inject:""`
+	Metrics        metrics.Metrics          `inject:"genericMetrics"`
+	SamplerFactory *sample.SamplerFactory   `inject:""`
+	SpanCache      cache.SpanCache          `inject:""`
+	// For test use only
+	BlockOnAddSpan bool
+
+	mut                   sync.RWMutex
+	samplersByDestination map[string]sample.Sampler
+
+	incoming chan *types.Span
+	reload   chan struct{}
+
+	eg *errgroup.Group
+}
+
+func (c *CentralCollector) Start() error {
+	// call reload config and then get the updated unique fields
+	collectorCfg, err := c.Config.GetCollectionConfig()
+	if err != nil {
+		return err
+	}
+
+	// listen for config reloads
+	c.Config.RegisterReloadCallback(c.sendReloadSignal)
+
+	c.incoming = make(chan *types.Span, collectorCfg.GetIncomingQueueSize())
+	c.reload = make(chan struct{}, 1)
+	c.samplersByDestination = make(map[string]sample.Sampler)
+
+	// spin up one collector because this is a single threaded collector
+	c.eg = &errgroup.Group{}
+	c.eg.Go(func() (err error) {
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("panic in central collector: %v\n%s", r, debug.Stack())
+			}
+		}()
+
+		c.collect()
+		return nil
+	})
+
+	return nil
+}
+
+func (c *CentralCollector) Stop() error {
+	close(c.incoming)
+	close(c.reload)
+	return c.eg.Wait()
+}
+
+// implement the Collector interface
+func (c *CentralCollector) AddSpan(span *types.Span) error {
+	return c.add(span, c.incoming)
+}
+
+func (c *CentralCollector) add(sp *types.Span, ch chan<- *types.Span) error {
+	if c.BlockOnAddSpan {
```

Comment:
We don't need the add function any more because we only have one span channel.

I'm also tempted to suggest we can drop BlockOnAddSpan, or at least should consider removing it once we have this integrated. Maybe we need it to make sure tests don't flake, but I'm not convinced we do anymore; I don't think any of our tests even approach the number of spans we'd need before this becomes a problem.